File: src/test/java/de/solidblocks/examples/javaspringwuphf/VisitCounterControllerTest.java
package de.solidblocks.examples.javaspringwuphf;

import static org.assertj.core.api.Assertions.assertThat;

import java.util.Map;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.NullAndEmptySource;
import org.junit.jupiter.params.provider.ValueSource;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;

class VisitCounterControllerTest {

    private static final StorageBackend BACKEND =
            new StorageBackend("memory", "In memory", "kept here", false, false, Map.of());

    private final VisitCounterController controller =
            new VisitCounterController(
                    new StubVisitCounter(), new ProcessEnvironment(Map.of("PATH", "/usr/bin")));

    private static class StubVisitCounter implements VisitCounter {
        private long visits;

        @Override
        public long recordVisit() {
            return ++visits;
        }

        @Override
        public StorageBackend storage() {
            return BACKEND;
        }
    }

    @Test
    void an_explicit_json_request_gets_the_report() {
        ResponseEntity<Object> response = controller.visit(MediaType.APPLICATION_JSON_VALUE);

        assertThat(response.getHeaders().getContentType()).isEqualTo(MediaType.APPLICATION_JSON);
        assertThat(response.getBody()).isInstanceOfSatisfying(
                VisitReport.class,
                report -> {
                    // The blcks integration tests read this number to know that the deployed
                    // application runs, thus it stays a number at the top of the document.
                    assertThat(report.visited()).isEqualTo(1);
                    assertThat(report.storage()).isEqualTo(BACKEND);
                    assertThat(report.environment().other()).isNotEmpty();
                });
    }

    @ParameterizedTest
    @NullAndEmptySource
    @ValueSource(
            strings = {
                "*/*",
                "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
                "text/html",
                "application"
            })
    void everything_else_gets_the_page(String accept) {
        ResponseEntity<Object> response = controller.visit(accept);

        assertThat(response.getHeaders().getContentType())
                .isEqualTo(new MediaType(MediaType.TEXT_HTML, java.nio.charset.StandardCharsets.UTF_8));
        assertThat(response.getBody()).isInstanceOfSatisfying(
                String.class, page -> assertThat(page).startsWith("<!doctype html>"));
    }

    @Test
    void a_json_request_that_also_takes_anything_else_gets_json() {
        assertThat(controller.visit("application/json, */*").getBody()).isInstanceOf(VisitReport.class);
    }

    @Test
    void every_request_counts_one_visit() {
        controller.visit(null);
        controller.visit(null);

        assertThat(((VisitReport) controller.visit(MediaType.APPLICATION_JSON_VALUE).getBody()).visited())
                .isEqualTo(3);
    }
}

File: src/test/kotlin/de/solidblocks/examples/javaspringwuphf/VisitCounterControllerTest.kt
package de.solidblocks.examples.javaspringwuphf

import io.kotest.matchers.collections.shouldNotBeEmpty
import io.kotest.matchers.shouldBe
import io.kotest.matchers.string.shouldStartWith
import io.kotest.matchers.types.shouldBeInstanceOf
import java.nio.charset.StandardCharsets
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.NullAndEmptySource
import org.junit.jupiter.params.provider.ValueSource
import org.springframework.http.MediaType

class VisitCounterControllerTest {
    private val controller =
        VisitCounterController(
            StubVisitCounter(),
            ProcessEnvironment(mapOf("PATH" to "/usr/bin")),
            wuphfPage(),
        )

    private class StubVisitCounter : VisitCounter {
        private var visits = 0L

        override fun recordVisit() = ++visits

        override fun storage() = BACKEND
    }

    @Test
    fun `an explicit json request gets the report`() {
        val response = controller.visit(MediaType.APPLICATION_JSON_VALUE)

        response.headers.contentType shouldBe MediaType.APPLICATION_JSON
        val report = response.body.shouldBeInstanceOf<VisitReport>()
        report.visited shouldBe 1
        report.storage shouldBe BACKEND
        report.environment.other.shouldNotBeEmpty()
    }

    @ParameterizedTest
    @NullAndEmptySource
    @ValueSource(
        strings =
            [
                "*/*",
                "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
                "text/html",
                "application",
            ]
    )
    fun `everything else gets the page`(accept: String?) {
        val response = controller.visit(accept)

        response.headers.contentType shouldBe MediaType(MediaType.TEXT_HTML, StandardCharsets.UTF_8)
        response.body.shouldBeInstanceOf<String>() shouldStartWith "<!doctype html>"
    }

    @Test
    fun `a json request that also takes anything else gets json`() {
        controller.visit("application/json, */*").body.shouldBeInstanceOf<VisitReport>()
    }

    @Test
    fun `every request counts one visit`() {
        controller.visit(null)
        controller.visit(null)

        controller.visit(MediaType.APPLICATION_JSON_VALUE).body.shouldBeInstanceOf<VisitReport>()
            .visited shouldBe 3
    }

    private companion object {
        val BACKEND = StorageBackend("memory", "In memory", "kept here", false, false, emptyMap())
    }
}
